crates/loader/src/validation.rs

```diff
@@ -1,23 +1,276 @@
+    } else {
+        parse_allowed_http_host_as_host(text)
+    }
+}
+
+fn parse_allowed_http_host_as_host(text: &str) -> Result<AllowedHttpHost, String> {
+    match text.split(':').collect::<Vec<_>>()[..] {
+        [host] => match is_valid_host_name(host) {
+            true => Ok(AllowedHttpHost::host(host)),
+            false => Err(format!("{} isn't a valid host or host:port string", text)),
+        },
+        [host, port] => match (is_valid_host_name(host), u16::from_str(port)) {
+            (true, Ok(p)) => {
+                if is_allowed_port(p) {
+                    Ok(AllowedHttpHost::host_and_port(host, p))
+                } else {
+                    Err(format!(
+                        "Can't allow {} because the port is reserved for non-HTTP protocols",
+                        text
+                    ))
+                }
+            }
+            _ => Err(format!("{} isn't a valid host or host:port string", text)),
+        },
+        _ => Err(format!("{} isn't a valid host or host:port string", text)),
+    }
+}
+
+fn is_valid_host_name(text: &str) -> bool {
+    return !text.is_empty()
+        && text.len() <= 253
+        && text
+            .chars()
+            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-' || c == '.');
+}
+
+fn parse_allowed_http_host_as_schemed_url(text: &str) -> Result<AllowedHttpHost, String> {
+    match Url::parse(text) {
+        Err(e) => Err(format!("{} isn't a valid HTTP host URL: {}", text, e)),
+        Ok(url) => {
+            match url.scheme() {
+                "http" | "https" => {
+                    let host = url.host_str();
+                    let port = url.port();
+                    let has_path = url.path().len() > 1; // allow "/"
+
+                    match (host, port.map(is_allowed_port).unwrap_or(true), has_path) {
+                        (None, _, _) => Err(format!("{} doesn't contain a host name", text)),
+                        (_, false, _) => Err(format!(
+                            "Can't allow {} because the port is reserved for non-HTTP protocols",
+                            text
+                        )),
+                        (_, _, true) => Err(format!(
+                            "{} contains a path, should be host and optional port only",
+                            text
+                        )),
+                        (Some(h), true, false) => Ok(AllowedHttpHost::new(h, port)),
+                    }
+                }
+                _ => Err(format!("{} isn't a valid host or host:port string", text)),
+            }
         }
-        let _ = domains
-            .iter()
-            .map(|d| {
-                Url::parse(d).with_context(|| format!("Can't parse {} in allowed_http_hosts", d))
-            })
-            .collect::<Result<Vec<_>, _>>()?;
-    }
-    Ok(())
+    }
+}
+
+fn is_allowed_port(port: u16) -> bool {
+    let forbidden_ports: &[u16] = &[25, 587, 2525];
```

Comment:
Top-level const? (or just drop this check per discussion w/ Mikkel)

Reply:
Yeah, I like Mikkel's suggestion.  I'll remove this and leave it to hosts to forbid these ports.